File: python/ch05/01-03.chatbot.py
from typing import Annotated, TypedDict

from langchain_core.messages import HumanMessage
from langchain_openai import ChatOpenAI
from langgraph.graph import StateGraph, START, END, add_messages
from langgraph.checkpoint.memory import MemorySaver


class State(TypedDict):
    # 메시지의 유형은 list이다. 
    # 어노테이션의 `add_messages` 함수는 상태를 업데이트하는 방법이다. 
    # 이 경우 이전 메시지를 대체하는 대신 새 메시지를 추가한다.
    messages: Annotated[list, add_messages]

builder = StateGraph(State)

model = ChatOpenAI(model='gpt-4o-mini')

def chatbot(state: State):
    answer = model.invoke(state['messages'])
    return {'messages': [answer]}

# 챗봇 노드 추가
# 첫 번째 인자는 고유한 노드 이름
# 두 번째 인자는 실행할 함수 또는 Runnable 
builder.add_node('chatbot', chatbot)

# 엣지 추가
builder.add_edge(START, 'chatbot')
builder.add_edge('chatbot', END)

graph = builder.compile()

# 그래프 이미지 저장
graph.get_graph().draw_mermaid_png(output_file_path='LLM_call_architecture.png')

# 그래프 실행
input = {'messages': [HumanMessage('안녕하세요!')]}
for chunk in graph.stream(input):
    print(chunk)

File: javascript/ch05/01-03.chatbot.js
import {
  StateGraph,
  Annotation,
  messagesStateReducer,
  START,
  END,
} from '@langchain/langgraph';
import { ChatOpenAI } from '@langchain/openai';
import { HumanMessage } from '@langchain/core/messages';
import * as fs from 'fs';

const State = {
  /**
   * State는 세 가지를 정의한다.
   * 1. 그래프 상태의 구조 (어떤 "채널"이 읽기/쓰기가 가능한지)
   * 2. 상태 채널의 기본값
   * 3. 상태 채널의 리듀서. 리듀서는 상태 업데이트 방법을 표현하는 함수를 말한다.
   *    아래에서는 새 메시지를 메시지 배열에 추가한다.
   */

  messages: Annotation({
    reducer: messagesStateReducer,
    default: () => [],
  }),
};

let builder = new StateGraph(State);

const model = new ChatOpenAI({model: 'gpt-4o-mini'});

async function chatbot(state) {
  const answer = await model.invoke(state.messages);
  return { messages: answer };
}

builder = builder.addNode('chatbot', chatbot);

builder = builder.addEdge(START, 'chatbot').addEdge('chatbot', END);

let graph = builder.compile();

// 그래프 이미지 저장
const image = await graph.getGraph().drawMermaidPng();
const arrayBuffer = await image.arrayBuffer();
const buffer = new Uint8Array(arrayBuffer);

fs.writeFileSync('LLM_call_architecture.png', buffer);

// 그래프 실행
const input = { messages: [new HumanMessage('안녕하세요!')] };
for await (const chunk of await graph.stream(input)) {
  console.log(chunk);
}
